python/ray/serve/_private/queue_monitor.py

```diff
@@ -0,0 +1,336 @@
+import logging
+from typing import Any, Dict
+
+import ray
+from ray.serve._private.constants import SERVE_LOGGER_NAME
+
+logger = logging.getLogger(SERVE_LOGGER_NAME)
+
+# Actor name prefix for QueueMonitor actors
+QUEUE_MONITOR_ACTOR_PREFIX = "QUEUE_MONITOR::"
+
+
+class QueueMonitorConfig:
+    """Configuration for the QueueMonitor deployment."""
+
+    def __init__(
+        self,
+        broker_url: str,
+        queue_name: str,
+    ):
+        self.broker_url = broker_url
+        self.queue_name = queue_name
+
+    @property
+    def broker_type(self) -> str:
+        url_lower = self.broker_url.lower()
+        if url_lower.startswith("redis"):
+            return "redis"
+        elif url_lower.startswith("amqp") or url_lower.startswith("pyamqp"):
+            return "rabbitmq"
+        else:
+            return "unknown"
+
+
+class QueueMonitor:
+
+    """
+    Actor that monitors queue length by directly querying the broker.
+
+    Returns pending tasks in the queue.
+
+    Uses native broker clients:
+        - Redis: Uses redis-py library with LLEN command
+        - RabbitMQ: Uses pika library with passive queue declaration
+    """
+
+    def __init__(self, config: QueueMonitorConfig):
+        self._config = config
+        self._last_queue_length: int = 0
+        self._is_initialized: bool = False
+
+        # Redis connection state
+        self._redis_client: Any = None
+
+        # RabbitMQ connection state
+        self._rabbitmq_connection: Any = None
+        self._rabbitmq_channel: Any = None
+
+    def initialize(self) -> None:
+        """
+        Initialize connection to the broker.
+
+        Creates the appropriate client based on broker type and tests the connection.
+        """
+        if self._is_initialized:
+            return
+
+        broker_type = self._config.broker_type
+        try:
+            if broker_type == "redis":
+                self._init_redis()
+            elif broker_type == "rabbitmq":
+                self._init_rabbitmq()
+            else:
+                raise ValueError(
+                    f"Unsupported broker type: {broker_type}. Supported: redis, rabbitmq"
+                )
+
+            self._is_initialized = True
+            logger.info(
+                f"QueueMonitor initialized for queue '{self._config.queue_name}' (broker: {broker_type})"
+            )
+
+        except Exception as e:
+            logger.error(f"Failed to initialize QueueMonitor: {e}")
+            raise
+
+    def _init_redis(self) -> None:
+        import redis
+
+        """Initialize Redis client."""
+        self._redis_client = redis.from_url(self._config.broker_url)
```

Comment:
does this make any assumption about authentication setting of redis? same question for rabbitmq